crypto/threads_pthread.c

```diff
@@ -42,6 +42,404 @@
+                                            int count)
+{
+    struct rcu_qp *new =
+        CRYPTO_zalloc(sizeof(struct rcu_qp) * count, NULL, 0);
+
+    ilock->group_count = count;
+    return new;
+}
+
+/*
+ * Public WRITE api
+ */
+void CRYPTO_THREAD_rcu_write_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_lock(&ilock->write_lock);
+}
+
+/*
+ * Public READ api
+ */
+void CRYPTO_THREAD_rcu_write_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_unlock(&ilock->write_lock);
+}
+
+/*
+ * Public WRITE api for synchronization
+ */
+void CRYPTO_THREAD_synchronize_rcu(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *ilock;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /*
+     * before we do anything else, lets grab the cb list
+     */
+    ilock = (struct rcu_lock_internal *)lock;
+    cb_items = __atomic_exchange_n(&ilock->cb_items, NULL, __ATOMIC_SEQ_CST);
+
+    qp = update_qp(ilock);
+
+    /*
+     * wait for the reader count to reach zero
+     */
+    do {
+        count = __atomic_load_n(&qp->users, __ATOMIC_SEQ_CST);
+    } while (READER_COUNT(count) != 0);
+
+    /*
+     * retire in order
+     */
+    pthread_mutex_lock(&ilock->prior_lock);
+    while (ilock->next_to_retire != ID_VAL(count))
+        pthread_cond_wait(&ilock->prior_signal, &ilock->prior_lock);
+    ilock->next_to_retire++;
+    pthread_cond_broadcast(&ilock->prior_signal);
+    pthread_mutex_unlock(&ilock->prior_lock);
+
+    retire_qp(ilock, qp);
+
+    /*
+     * handle any callbacks that we have
+     */
+    while (cb_items != NULL) {
+        tmpcb = cb_items;
+        cb_items = cb_items->next;
+        tmpcb->fn(tmpcb->data);
+        CRYPTO_free(tmpcb, NULL, 0);
+    }
+    /*
+     * and we're done
+     */
+    return;
+}
+
+/*
+ * Public WRITE api to schedule a callback after a
+ * completed synchronization
+ */
+void CRYPTO_THREAD_rcu_call(CRYPTO_RCU_LOCK *lock, rcu_cb_fn cb, void *data)
+{
+    struct rcu_cb_item *new =
+        CRYPTO_zalloc(sizeof(struct rcu_cb_item), NULL, 0);
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    new->data = data;
+    new->fn = cb;
+    new->next = __atomic_exchange_n(&ilock->cb_items, new, __ATOMIC_SEQ_CST);
+}
+
+/*
+ * Public READ/WRITE api to get the value of a pointer
+ */
+void *CRYPTO_THREAD_rcu_uptr_derefrence(uintptr_t *p)
+{
+    return (void *)__atomic_load_n(p, __ATOMIC_ACQUIRE);
+}
+
+/*
+ * Public WRITE api to assign a pointer value
+ */
+void CRYPTO_THREAD_rcu_assign_uptr(uintptr_t *p, uintptr_t *v)
+{
+    __atomic_store(p, v, __ATOMIC_RELEASE);
+}
+
+/*
+ * Public api to allocate a new rcu lock
+ */
+CRYPTO_RCU_LOCK *CRYPTO_THREAD_rcu_lock_new(void)
+{
+    struct rcu_lock_internal *new =
+        CRYPTO_zalloc(sizeof(struct rcu_lock_internal), NULL, 0);
```

Comment:
OPENSSL_zalloc?

Reply:
Its a crypto api, I would have thought the CRYPTO_*alloc api would have been the appropriate one to use.  But sure, I can change it